vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,476 @@
+            Some(invalid_gpa) => return Err(DiagnosticsError::BadGpa(invalid_gpa)),
+            None => {
+                return Err(DiagnosticsError::GeneralError(
+                    "No diagnostics GPA set".to_string(),
+                ));
+            }
+        };
+
+        //
+        // Step 2: Read and validate the advanced logger header
+        //
+        let header: AdvancedLoggerInfo = gm.read_plain(gpa as u64)?;
+        header.validate()?;
+
+        //
+        // Step 3: Prepare processing variables
+        //
+
+        // Force clear the logs
+        logs.clear();
+
+        // Used for summary statistics
+        let mut bytes_read: usize = 0;
+        let mut entries_processed: usize = 0;
+
+        // Copy packed fields to local variables to avoid unaligned access
+        let log_current_offset = header.log_current_offset;
+        let log_buffer_offset = header.log_buffer_offset;
+
+        // Calculate used log buffer size using the local variables
+        let used_log_buffer_size = log_current_offset
+            .checked_sub(log_buffer_offset)
+            .ok_or_else(|| {
+                DiagnosticsError::Overflow(format!(
+                    "log_current_offset ({:#x}) - log_buffer_offset ({:#x})",
+                    log_current_offset, log_buffer_offset
+                ))
+            })?;
+
+        // Validate used log buffer size
+        if used_log_buffer_size == 0
+            || used_log_buffer_size > header.log_buffer_size
+            || used_log_buffer_size > MAX_LOG_BUFFER_SIZE
+        {
+            return Err(DiagnosticsError::GeneralError(format!(
+                "Invalid used log buffer size: {:#x}, max: {:#x}",
+                used_log_buffer_size, MAX_LOG_BUFFER_SIZE
+            )));
+        }
+
+        // Used for accumulating multiple messages
+        let mut accumulated_message = String::with_capacity(MAX_MESSAGE_LENGTH as usize);
+        let mut debug_level = 0;
+        let mut time_stamp = 0;
+        let mut phase = 0;
+        let mut is_accumulating = false;
+
+        //
+        // Step 4: Read the used portions of the log buffer
+        //
+
+        // Calculate start address of the log buffer
+        let buffer_start_addr = gpa.checked_add(log_buffer_offset).ok_or_else(|| {
+            DiagnosticsError::Overflow(format!(
+                "gpa ({:#x}) + log_buffer_offset ({:#x})",
+                gpa, log_buffer_offset
+            ))
+        })?;
+
+        let mut buffer_data = vec![0u8; used_log_buffer_size as usize];
+        gm.read_at(buffer_start_addr as u64, &mut buffer_data)?;
+
+        // Empty buffer data should early exit
+        if buffer_data.is_empty() {
+            tracelimit::info_ratelimited!("buffer_data is empty, ending processing");
+            return Ok(());
+        }
+
+        //
+        // Step 5: Parse the log buffer
+        //
+        let mut buffer_slice = &buffer_data[..];
+        while !buffer_slice.is_empty() {
```

Comment:
I suggest splitting out the code that parses an entry to a separate function. Something like:

```rust
struct Entry<'a> {
    debug_level: X,
    message: &'a str,
    // etc.
}
fn parse_entry<'a>(data: &mut &'a [u8]) -> Result<Entry<'a>> { ... }
```

Then this loop gets a lot simpler and easier to review, because it's just responsible for accumulating messages (which you can probably do more cleanly with a nested loop, at that point, eliminating the need for a bunch of `mut`s).